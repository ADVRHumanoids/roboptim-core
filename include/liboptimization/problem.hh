// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the Problem class.
 */

#ifndef OPTIMIZATION_PROBLEM_HH
# define OPTIMIZATION_PROBLEM_HH
# include <boost/optional.hpp>
# include <boost/static_assert.hpp>
# include <boost/type_traits/is_convertible.hpp>

# include <liboptimization/fwd.hh>
# include <liboptimization/function.hh>

namespace optimization
{
  /// Define a problem in a generic way.
  template <typename F, typename C>
  class Problem
  {
    BOOST_STATIC_ASSERT((boost::is_convertible<F*, Function*>::value));
  public:
    typedef F function_t;
    typedef C constraint_t;

    typedef std::vector<constraint_t> constraints_t;
    typedef boost::optional<Function::vector_t> startingPoint_t;

    explicit Problem (const function_t&) throw ();

    /// Copy constructor.
    explicit Problem (const Problem<F, C>&) throw ();

    /// Copy constructor (convert from another class of problem).
    template <typename F_, typename C_>
    explicit Problem (const Problem<F_, C_>&) throw ();

    ~Problem () throw ();

    const function_t& function () const throw ();

    const constraints_t& constraints () const throw ();
    constraints_t& constraints () throw ();

    startingPoint_t& startingPoint () throw ();
    const startingPoint_t& startingPoint () const throw ();

  private:
    const function_t& function_;
    startingPoint_t startingPoint_;
    constraints_t constraints_;
  };
}; // end of namespace optimization

# include <liboptimization/problem.hxx>
#endif //! OPTIMIZATION_PROBLEM_HH
