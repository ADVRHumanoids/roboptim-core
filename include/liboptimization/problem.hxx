// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Implementation of the Problem class.
 */

#ifndef OPTIMIZATION_PROBLEM_HXX
# define OPTIMIZATION_PROBLEM_HXX
# include <algorithm>
# include <boost/type_traits/remove_pointer.hpp>

namespace optimization
{
  template <typename F, typename C>
  Problem<F, C>::Problem (const function_t& f) throw ()
    : function_ (f),
      startingPoint_ (),
      constraints_ ()
  {
  }

  template <typename F, typename C>
  Problem<F, C>::~Problem () throw ()
  {
  }

  // Copy constructor.
  template <typename F, typename C>
  Problem<F, C>::Problem (const Problem<F, C>& pb) throw ()
    : function_ (pb.function_),
      startingPoint_ (pb.startingPoint_),
      constraints_ (pb.constraints_)
  {
  }

  // Copy constructor (convert from another class of problem).
  template <typename F, typename C>
  template <typename F_, typename C_>
  Problem<F, C>::Problem (const Problem<F_, C_>& pb) throw ()
    : function_ (pb.function_),
      startingPoint_ (pb.startingPoint_),
      constraints_ ()
  {
    // Check that F is a subtype of F_.
    BOOST_STATIC_ASSERT((boost::is_convertible<F_*, F*>::value));
    // Check that C is a subtype of C_.

    typedef typename boost::remove_pointer<C_>::type rpC_;
    typedef typename boost::remove_pointer<C>::type rpC;
    BOOST_STATIC_ASSERT((boost::is_convertible<rpC_*, rpC*>::value));

    std::copy (pb.constraints_.begin (), pb.constraints_.end (),
               constraints_.begin ());
  }

  template <typename F, typename C>
  const typename Problem<F, C>::function_t&
  Problem<F, C>::function () const throw ()
  {
    return function_;
  }

  template <typename F, typename C>
  typename Problem<F, C>::constraints_t&
  Problem<F, C>::constraints () throw ()
  {
    return constraints_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::constraints_t&
  Problem<F, C>::constraints () const throw ()
  {
    return constraints_;
  }

  template <typename F, typename C>
  typename Problem<F, C>::startingPoint_t&
  Problem<F, C>::startingPoint () throw ()
  {
    return startingPoint_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::startingPoint_t&
  Problem<F, C>::startingPoint () const throw ()
  {
    return startingPoint_;
  }

}; // end of namespace optimization

# include <liboptimization/problem.hxx>
#endif //! OPTIMIZATION_PROBLEM_HH
