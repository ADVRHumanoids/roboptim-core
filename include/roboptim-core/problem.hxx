// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Implementation of the Problem class.
 */

#ifndef ROBOPTIM_CORE_PROBLEM_HXX
# define ROBOPTIM_CORE_PROBLEM_HXX
# include <algorithm>
# include <boost/numeric/ublas/io.hpp>
# include <boost/type_traits/is_pointer.hpp>
# include <boost/type_traits/remove_pointer.hpp>
# include <boost/variant.hpp>
# include <boost/variant/get.hpp>
# include <boost/variant/apply_visitor.hpp>

# include <roboptim-core/indent.hh>
# include <roboptim-core/util.hh>

namespace roboptim
{
  /**
     \addtogroup roboptim_problem
     @{
  */

  template <typename F, typename C>
  Problem<F, C>::Problem (const function_t& f) throw ()
    : function_ (f),
      startingPoint_ (),
      constraints_ (),
      bounds_ (),
      argBounds_ (f.n),
      scales_ (),
      argScales_ (f.n)
  {
    // Check that in the objective function m = 1 (R^n -> R).
    assert (f.m == 1);

    // Initialize bound.
    std::fill (argBounds_.begin (), argBounds_.end (), makeInfiniteBound ());
    // Initialize scale.
    std::fill (argScales_.begin (), argScales_.end (), 1.);
  }

  template <typename F, typename C>
  Problem<F, C>::~Problem () throw ()
  {
  }

  // Copy constructor.
  template <typename F, typename C>
  Problem<F, C>::Problem (const Problem<F, C>& pb) throw ()
    : function_ (pb.function_),
      startingPoint_ (pb.startingPoint_),
      constraints_ (pb.constraints_),
      bounds_ (pb.bounds_),
      argBounds_ (pb.argBounds_),
      scales_ (pb.scales_),
      argScales_ (pb.argScales_)
  {
  }

  // Copy constructor (convert from another class of problem).
  template <typename F, typename C>
  template <typename F_, typename C_>
  Problem<F, C>::Problem (const Problem<F_, C_>& pb) throw ()
    : function_ (pb.function_),
      startingPoint_ (pb.startingPoint_),
      constraints_ (),
      bounds_ (pb.bounds_),
      argBounds_ (pb.argBounds_),
      scales_ (pb.scales_),
      argScales_ (pb.argScales_)
  {
    // Check that F is a subtype of F_.
    BOOST_STATIC_ASSERT((boost::is_base_of<F, F_>::value));

    // Check that C is a subtype of C_.
    typedef typename boost::remove_pointer<C_>::type rpC_;
    typedef typename boost::remove_pointer<C>::type rpC;
    BOOST_STATIC_ASSERT((boost::is_base_of<rpC, rpC_>::value));

    std::copy (pb.constraints_.begin (), pb.constraints_.end (),
               constraints_.begin ());
  }

  template <typename F, typename C>
  const typename Problem<F, C>::function_t&
  Problem<F, C>::function () const throw ()
  {
    return function_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::constraints_t&
  Problem<F, C>::constraints () const throw ()
  {
    return constraints_;
  }

  template <typename F, typename C>
  void
  Problem<F, C>::addConstraint (const C& x, value_type s)
    throw (std::runtime_error)
  {
    addConstraint (x, makeInfiniteBound (), s);
  }

  template <typename F, typename C>
  void
  Problem<F, C>::addConstraint (const C& x, bound_t b, value_type s)
    throw (std::runtime_error)
  {
    constraints_.push_back (x);
    bounds_.push_back (b);
    scales_.push_back (s);
  }

  template <typename F, typename C>
  typename Problem<F, C>::startingPoint_t&
  Problem<F, C>::startingPoint () throw ()
  {
    return startingPoint_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::startingPoint_t&
  Problem<F, C>::startingPoint () const throw ()
  {
    return startingPoint_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::bounds_t&
  Problem<F, C>::bounds () const throw ()
  {
    return bounds_;
  }

  template <typename F, typename C>
  typename Problem<F, C>::bounds_t&
  Problem<F, C>::argBounds () throw ()
  {
    return argBounds_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::bounds_t&
  Problem<F, C>::argBounds () const throw ()
  {
    return argBounds_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::scales_t&
  Problem<F, C>::scales () const throw ()
  {
    return scales_;
  }

  template <typename F, typename C>
  typename Problem<F, C>::scales_t&
  Problem<F, C>::argScales () throw ()
  {
    return argScales_;
  }

  template <typename F, typename C>
  const typename Problem<F, C>::scales_t&
  Problem<F, C>::argScales () const throw ()
  {
    return argScales_;
  }


  namespace detail
  {
    template <typename T>
    std::ostream&
    impl_print (std::ostream& o, const T* t)
    {
      return o << *t;
    }

    template <typename T>
    std::ostream&
    impl_print (std::ostream& o, const T& t)
    {
      return o << t;
    }
  }

  template <typename F, typename C>
  std::ostream&
  Problem<F, C>::print (std::ostream& o) const throw ()
  {
    o << "Problem:" << incendl;
    // Function.
    o << function () << iendl;

    // Arguments' bounds.
    o << "Argument's bounds: " << argBounds () << iendl;
    // Arguments' scales.
    o << "Argument's scales: " << argScales () << iendl;

    // Constraints.
    if (constraints ().empty ())
      o << "No constraints.";
    else
      o << "Number of constraints: " << constraints ().size ();

    typedef typename constraints_t::const_iterator citer_t;
    unsigned i = 0;
    for (citer_t it = constraints ().begin ();
         it != constraints ().end ();
         ++it, ++i)
      {
        o << iendl << incindent
          << "Constraint " << i << incindent << iendl
          << "Bounds: " << bounds ()[i] << iendl
          << "Scales: " << scales ()[i] << iendl
          << decindent << decindent;
      }

    // Starting point.
    if (startingPoint_)
      o << iendl << "Starting point: " << *startingPoint_;
    else
      o << iendl << "No starting point.";

    // Infinity.
    o << iendl << "Infinity value (for all functions): " << Function::infinity ();

    // Arguments' bounds.
    // Arguments' scales.


    return o << decindent;
  }

  template <typename F, typename C>
  std::ostream&
  operator<< (std::ostream& o, const Problem<F, C>& pb)
  {
    return pb.print (o);
  }
  /**
     @}
  */
}; // end of namespace roboptim
# include <roboptim-core/problem.hxx>
#endif //! ROBOPTIM_CORE_PROBLEM_HH
