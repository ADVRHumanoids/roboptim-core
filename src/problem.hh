// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \file src/problem.hh
 *
 * \brief Declaration of the Problem class.
 */

#ifndef OPTIMIZATION_PROBLEM_HH
# define OPTIMIZATION_PROBLEM_HH
# include <vector>
# include <boost/shared_ptr.hpp>

# include <fwd.hh>
# include <function.hh>

namespace optimization
{
  /// Define an optimization problem.
  struct Problem
  {
    /// Smart pointer for function arrays.
    typedef boost::shared_ptr<Function> functionPtr_t;
    /// Vector of functions which symbolizes constraints.
    typedef std::vector<functionPtr_t> constraints_t;

    /// Constructor.
    Problem (const Function&) throw ();

    /// Destructor.
    ~Problem () throw ();

    /// Function which has to be minimized.
    const Function& function;

    /// Constraints.
    /// Fast to compute constraints should be first
    /// (as they are usually evaluted first).
    constraints_t constraints;

    /// Starting point.
    Function::vector_t start;
  };
} // end of namespace optimization

#endif //! OPTIMIZATION_PROBLEM_HH
