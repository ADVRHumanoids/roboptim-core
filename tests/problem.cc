// Copyright (C) 2014 by Benjamin Chr√©tien, CNRS.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <roboptim/core/io.hh>
#include <roboptim/core/problem.hh>
#include <roboptim/core/function/constant.hh>

using namespace roboptim;

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE (problem)
{
  typedef Problem<DifferentiableFunction,
		  boost::mpl::vector<DifferentiableFunction> >
    problem_t;
  typedef Function::argument_t argument_t;
  typedef problem_t::startingPoint_t startingPoint_t;

  ConstantFunction::vector_t v (2);
  v.setZero ();

  ConstantFunction f (v);

  problem_t pb_fail (f);

  argument_t bad_x (3);
  pb_fail.startingPoint () = bad_x;

  {
    BOOST_CHECK_THROW (startingPoint_t test_sp = pb_fail.startingPoint ();
                       std::cout << (*test_sp) << std::endl,
                       std::runtime_error);
  }

  {
    BOOST_CHECK_THROW (const startingPoint_t& test_sp = pb_fail.startingPoint ();
                       std::cout << (*test_sp) << std::endl,
                       std::runtime_error);
  }

  problem_t pb (f);
  argument_t x (2);
  x.setZero ();
  pb.startingPoint () = x;

  boost::shared_ptr<ConstantFunction>
    cstr = boost::make_shared<ConstantFunction>  (v);
  problem_t::intervals_t intervals (2);
  problem_t::scales_t scales (2, 1);
  for (size_t i = 0; i < intervals.size (); ++i)
    intervals[i] = Function::makeInfiniteInterval ();
  pb.addConstraint (cstr, intervals, scales);

  // Check null ptr
  BOOST_CHECK_THROW (boost::shared_ptr<ConstantFunction> null_ptr;
                     pb.addConstraint (null_ptr, intervals, scales),
                     std::runtime_error);

  // Check invalid input size
  BOOST_CHECK_THROW (ConstantFunction::vector_t v_size (4);
                     boost::shared_ptr<ConstantFunction>
                     cstr_size = boost::make_shared<ConstantFunction> (v_size);
                     pb.addConstraint (cstr_size, intervals, scales),
                     std::runtime_error);

  // Check invalid interval size
  BOOST_CHECK_THROW (problem_t::intervals_t intervals_size (6);
                     pb.addConstraint (cstr, intervals_size, scales),
                     std::runtime_error);

  // Check invalid scale vector size
  BOOST_CHECK_THROW (problem_t::scales_t scales_size (6);
                     pb.addConstraint (cstr, intervals, scales_size),
                     std::runtime_error);

  std::cout << pb << std::endl;
}

BOOST_AUTO_TEST_SUITE_END ()
