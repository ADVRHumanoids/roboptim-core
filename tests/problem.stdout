[4,2]((0,0), (0,0), (1,0), (0,1))
Problem:
  Constant function
    offset = [2](0,0)
  Arguments bounds: (-5, 5), (-10, inf)
  Arguments scaling: 1, 1
  Arguments names: x₀, x₁
  Number of constraints: 3
  Constraint 0
      Constant function
        offset = [2](3,-1)
      Bounds: (-inf, inf), (-inf, inf)
      Scaling: 1, 1
      Initial value: [2](3, -1)
  Constraint 1
      Numeric linear function:
        A = [2,2]((1,0), (0,1))
        B = [2](2,3)
      Bounds: (0, inf), (0, inf)
      Scaling: 1, 1
      Initial value: [2](2, 3)
  Constraint 2
      a * b, a + b (not differentiable)
      Bounds: (-12, 42), (-12, 42)
      Scaling: 1, 1
      Initial value: [2](0, 0)
  Starting point: [2](0,0)
  Starting value: [2](0,0)
  Infinity value (for all functions): inf
Problem:
  Constant function
    offset = [2](3,-1)
  Arguments bounds: (-inf, inf), (-inf, inf)
  Arguments scaling: 1, 1
  No constraints.
  No starting point.
  Infinity value (for all functions): inf
Problem:
  Constant function
    offset = [2](0,0)
  Arguments bounds: (-5, 5), (-10, inf)
  Arguments scaling: 1, 1
  Arguments names: x₀, x₁
  No constraints.
  Starting point: [2](0,0)
  Starting value: [2](0,0)
  Infinity value (for all functions): inf
Problem:
  Constant function
    offset = [2](0,0)
  Arguments bounds: (-5, 5), (-10, inf)
  Arguments scaling: 1, 1
  Arguments names: x₀, x₁
  No constraints.
  Starting point: [2](200,30)
  Starting value: [2](0,0)
  Infinity value (for all functions): inf
Problem:
  Constant function
    offset = [2](0,0)
  Arguments bounds: (-inf, 5), (-inf, inf)
  Arguments scaling: 1, 1
  Arguments names: x₀, x₁
  Number of constraints: 2
  Constraint 0
      Constant function
        offset = [2](3,-1)
      Bounds: (-12, 42), (-12, 42)
      Scaling: 1, 1
      Initial value: [2](3, -1)
  Constraint 1
      Numeric linear function:
        A = [2,2]((1,0), (0,1))
        B = [2](2,3)
      Bounds: (-12, 42), (-12, 42)
      Scaling: 1, 1
      Initial value: [2](2, 3)
  Starting point: [2](0,0)
  Starting value: [2](0,0)
  Infinity value (for all functions): inf
Problem:
  Constant function
    offset = [2](0,0)
  Arguments bounds: (-inf, 5), (-inf, inf)
  Arguments scaling: 1, 1
  Arguments names: x₀, x₁
  Number of constraints: 2
  Constraint 0
      Constant function
        offset = [2](3,-1)
      Bounds: (-12, 42), (-12, 42)
      Scaling: 1, 1
      Initial value: [2](3, -1)
  Constraint 1
      Numeric linear function:
        A = [2,2]((1,0), (0,1))
        B = [2](2,3)
      Bounds: (-12, 42), (-12, 42)
      Scaling: 1, 1
      Initial value: [2](202, 53) (constraint not satisfied)
  Starting point: [2](200,50)
  Starting value: [2](0,0)
  Infinity value (for all functions): inf
